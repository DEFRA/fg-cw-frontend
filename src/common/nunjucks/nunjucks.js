import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
