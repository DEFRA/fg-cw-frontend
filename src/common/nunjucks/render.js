import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
