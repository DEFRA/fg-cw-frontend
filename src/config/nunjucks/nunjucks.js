import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
