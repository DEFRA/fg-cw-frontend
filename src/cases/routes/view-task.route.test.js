import hapi from "@hapi/hapi";
import Yar from "@hapi/yar";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { createMockLinks } from "../../../test/data/case-test-data.js";
import { nunjucks } from "../../common/nunjucks/nunjucks.js";
import { findCaseByIdUseCase } from "../use-cases/find-case-by-id.use-case.js";
import { viewTaskRoute } from "./view-task.route.js";

vi.mock("../use-cases/find-case-by-id.use-case.js");

describe("viewTaskRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    await server.register({
      plugin: Yar,
      options: {
        name: "session",
        cookieOptions: {
          password: "abcdefghijklmnopqrstuvwxyz012345",
          isSecure: false,
          isSameSite: "Strict",
        },
      },
    });

    server.route(viewTaskRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("returns a task", async () => {
    findCaseByIdUseCase.mockResolvedValue({
      _id: "68495db5afe2d27b09b2ee47",
      caseRef: "banana-123",
      workflowCode: "frps-private-beta",
      status: "NEW",
      dateReceived: "2025-06-11T10:43:01.603Z",
      currentStage: "application-receipt",
      links: createMockLinks("68495db5afe2d27b09b2ee47"),
      payload: {
        clientRef: "banana-123",
        code: "frps-private-beta",
        createdAt: "2025-06-11T10:43:01.417Z",
        submittedAt: "2023-10-01T12:00:00.000Z",
        identifiers: {
          sbi: "SBI001",
          frn: "FIRM0001",
          crn: "CUST0001",
          defraId: "DEFRA0001",
        },
        answers: {
          agreementName: "Test application name 1",
          scheme: "SFI",
          year: 2025,
          hasCheckedLandIsUpToDate: true,
          actionApplications: [
            {
              parcelId: "9238",
              sheetId: "SX0679",
              code: "CSAM1",
              appliedFor: {
                unit: "ha",
                quantity: 20.23,
              },
            },
          ],
        },
      },
      stages: [
        {
          code: "application-receipt",
          name: "Application Receipt",
          taskGroups: [
            {
              code: "application-receipt-tasks",
              title: "Application Receipt Tasks",
              tasks: [
                {
                  id: "simple-review",
                  title: "Simple Review",
                  status: "pending",
                  type: "OPTIONAL",
                },
              ],
            },
          ],
        },
        {
          code: "contract",
          name: "Contract",
          taskGroups: [],
        },
      ],
      comments: [],
    });

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/tasks/application-receipt-tasks/simple-review",
      auth: {
        credentials: { token: "mock-token" },
        strategy: "session",
      },
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });
});
