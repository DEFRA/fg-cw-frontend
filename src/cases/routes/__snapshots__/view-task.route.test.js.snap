// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viewTaskRoute > returns a task 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>

  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Tasks</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="#">
                                    
Notes
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>


    

<div class="govuk-grid-row">
  <div class="govuk-grid-column-two-thirds">
    <a href="/cases/68495db5afe2d27b09b2ee47" class="govuk-back-link" data-testid="back-link">Back to task list</a>

    <h2 class="govuk-heading-l govuk-!-margin-top-6" data-testid="task-details-heading">
      Simple Review
    </h2>

  </div>
</div>

  </div>
        "
`;
